### PR TITLE
Remove usage of internal header encoding API

### PR DESCRIPTION
**Pull request by MadLittleMods**
Remove usage of internal header encoding API

```py
from twisted.web.http_headers import Headers

Headers()._canonicalNameCaps
Headers()._encodeName
```

Introduced in https://github.com/matrix-org/synapse/pull/15913 <- https://github.com/matrix-org/synapse/pull/15773

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] [Code style](https://element-hq.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: tests/http/test_proxy.py [hunk 1]
```diff
@@ -30,19 +30,19 @@
...
             # Normalizes header capitalization
-            [b"keep-alive, x-fOo, x-bAr", {"Keep-Alive", "X-Foo", "X-Bar"}],
+            [b"keep-alive, x-fOo, x-bAr", {"keep-alive", "x-foo", "x-bar"}],
             # Handles header names with whitespace
             [
                 b"keep-alive, x  foo, x bar",
-                {"Keep-Alive", "X  foo", "X bar"},
+                {"keep-alive", "x  foo", "x bar"},
             ],
         ]
```
Comment (erikjohnston) [inline tests/http/test_proxy.py:47]: Looking at the twisted code they do some special casing of specific headers. I think we're fine, but can we add a test to ensure that we correctly match everything in `HOP_BY_HOP_HEADERS_LOWERCASE`?

Comment (MadLittleMods) [reply tests/http/test_proxy.py:47]: > Looking at the twisted code they do some special casing of specific headers.

We should get this same behavior after passing it to `Headers.setRawHeaders(...)` which uses the same `_encodeName(...)` that we were using before to do that special casing.


> I think we're fine, but can we add a test to ensure that we correctly match everything in `HOP_BY_HOP_HEADERS_LOWERCASE`?

Can update :fast_forward:

File: tests/http/test_matrixfederationclient.py [hunk 2]
```diff
@@ -903,9 +903,10 @@ def test_proxy_requests_and_discards_hop_by_hop_headers(self) -> None:
                     headers=Headers(
                         {
                             "Content-Type": ["application/json"],
+                            "X-Test": ["test"],
                             # Define some hop-by-hop headers (try with varying casing to
                             # make sure we still match-up the headers)
-                            "Connection": ["close, X-fOo, X-Bar", "X-baz"],
+                            "Connection": ["close, X-fOo, X-Bar, X-baz"],
```
Comment (erikjohnston) [inline tests/http/test_matrixfederationclient.py:909]: Why are we doing this? I think this was testing what happens if we have multiple `Connection` headers in the request (which I think is valid)

Comment (MadLittleMods) [reply tests/http/test_matrixfederationclient.py:909]: It was a mistake within this PR (you're commenting on a commit diff as far as I can tell). On `develop`, we weren't testing this before at all and we only take into account the first `Connection` header (according to the code, `connection_header[0]`).

I added `X-baz` to test the "not in canonical case"  case but the assertion never checked it. The last commit fixes all of this.

**Review by erikjohnston — APPROVED** [review]

**Comment by MadLittleMods** [comment]
Thanks for the review @erikjohnston 🐦

**Pull request merged**